### PR TITLE
Add mana to balance

### LINKED ISSUE #1865: Add mana to Balance

### PR DESCRIPTION
**Pull request by DaughterOfMars**
# Description of change

Adds potential mana and stored mana (both with decay) to the balance calculation.

## Links to any relevant issues

Closes #1865

### CODE REVIEW

File: sdk/src/wallet/types/balance.rs
Comment (Thoralf-M) [inline sdk/src/wallet/types/balance.rs]: Maybe something like this
```suggestion
    /// Decayed potential Mana of owned outputs.
    pub(crate) potential_mana: u64,
    /// Decayed stored Mana of owned outputs.
    pub(crate) stored_mana: u64,
    /// Block issuance credits of wallet accounts.
    pub(crate) block_issuance_credits: BTreeMap<AccountId, i128>,
```
also compared to the base coin, these values would add up to the `total Mana`, are we maybe also interested in `available Mana`?

Comment (thibault-martinez) [reply sdk/src/wallet/types/balance.rs]: Yeah I'd maybe even put that in a new struct like BaseCoin

File: sdk/src/wallet/operations/balance.rs
Comment (Thoralf-M) [inline sdk/src/wallet/operations/balance.rs]: The required amount for the storage deposit needs to be subtracted here and in the other places

File: sdk/src/wallet/operations/balance.rs
Comment (Thoralf-M) [inline sdk/src/wallet/operations/balance.rs]: The requests should be sent in parallel so it doesn't really take longer with multiple accounts

File: sdk/src/wallet/operations/balance.rs
Comment (DaughterOfMars) [reply sdk/src/wallet/operations/balance.rs]: Are you sure? Isn't that still generating mana for this wallet?

File: sdk/src/wallet/operations/balance.rs
Comment (DaughterOfMars) [reply sdk/src/wallet/operations/balance.rs]: tbh I think we shouldn't call this here anyway, since we're holding the wallet data hostage during this entire process.

Comment (Thoralf-M) [reply sdk/src/wallet/operations/balance.rs]: Why not move it to the beginning before then?

Comment (Thoralf-M) [reply sdk/src/wallet/operations/balance.rs]: Not sure about caching since the BIC also gets decayed all the time

File: sdk/src/wallet/operations/balance.rs
Comment (Thoralf-M) [reply sdk/src/wallet/operations/balance.rs]: Yes `Every IOTA coin (besides storage deposits) generates potential Mana.` https://github.com/iotaledger/tips/blob/tip39/tips/TIP-0039/tip-0039.md#potential-mana

Comment (DaughterOfMars) [reply sdk/src/wallet/operations/balance.rs]: ahh

File: sdk/src/wallet/types/balance.rs
Comment (kwek20) [inline sdk/src/wallet/types/balance.rs]: Cant we derive this with derive_more?

Comment (DaughterOfMars) [reply sdk/src/wallet/types/balance.rs]: tru

**Review by Thoralf-M — DISMISSED** [review]

File: sdk/src/wallet/types/balance.rs
Comment (thibault-martinez) [inline sdk/src/wallet/types/balance.rs]:
```suggestion
/// Mana fields for [`Balance`]
```

File: sdk/src/wallet/types/balance.rs
Comment (thibault-martinez) [inline sdk/src/wallet/types/balance.rs]:
```suggestion
    /// Total and available mana
```

File: sdk/src/wallet/types/balance.rs
Comment (thibault-martinez) [reply sdk/src/wallet/types/balance.rs]: But then it's the same comment as the other one so do something about it :trollface:

**Comment by Thoralf-M** [comment]
Another remainder https://github.com/iotaledger/iota-sdk/pull/1874#discussion_r1464915386

File: sdk/src/wallet/types/balance.rs
Comment (DaughterOfMars) [reply sdk/src/wallet/types/balance.rs]: oop

Comment (DaughterOfMars) [reply sdk/src/wallet/types/balance.rs]: I think I'll wait until #1421 is merged just so I can use that new struct

Comment (DaughterOfMars) [reply sdk/src/wallet/types/balance.rs]: wait nvm it's a different pr and it's already merged lol

File: sdk/src/types/block/output/mod.rs
Comment (thibault-martinez) [inline sdk/src/types/block/output/mod.rs:448]: There are getters!!!

Comment (DaughterOfMars) [reply sdk/src/types/block/output/mod.rs:448]: oh oops lol

Comment (DaughterOfMars) [reply sdk/src/types/block/output/mod.rs:448]: wait no I need mutable access

Comment (thibault-martinez) [reply sdk/src/types/block/output/mod.rs:448]: Ah, my bad then

File: sdk/src/wallet/operations/balance.rs
Comment (thibault-martinez) [inline sdk/src/wallet/operations/balance.rs:71]: Can you not just call `decayed_mana` on the output then?

Comment (DaughterOfMars) [reply sdk/src/wallet/operations/balance.rs:71]: yeah right I forgot those methods were added

Comment (DaughterOfMars) [reply sdk/src/wallet/operations/balance.rs:71]: done

**Review by Thoralf-M — APPROVED** [review]

**Review by thibault-martinez — APPROVED** [review]

**Pull request merged**
